chainstate/tx-verifier/src/transaction_verifier/mod.rs

```diff
@@ -157,81 +157,118 @@ pub struct TransactionVerifierDelta {
     token_issuance_cache: ConsumedTokenIssuanceCache,
 }
 
+/// [`TxIndexCache`] that can be enabled or disabled (using a config).
+pub struct GuardedTxIndexCache {
+    enabled: bool,
+    inner: TxIndexCache,
+}
+
+impl GuardedTxIndexCache {
+    fn new(enabled: bool) -> Self {
+        let inner = TxIndexCache::new();
+        Self { enabled, inner }
+    }
+
+    fn from_config(config: &TransactionVerifierConfig) -> Self {
+        Self::new(config.tx_index_enabled)
+    }
+
+    #[cfg(test)]
+    fn new_for_test(map: BTreeMap<OutPointSourceId, CachedInputsOperation>) -> Self {
+        let inner = TxIndexCache::new_for_test(map);
+        let enabled = true;
+        Self { enabled, inner }
+    }
+
+    fn as_ref(&self) -> Option<&TxIndexCache> {
+        self.enabled.then_some(&self.inner)
+    }
+
+    fn as_mut(&mut self) -> Option<&mut TxIndexCache> {
+        self.enabled.then_some(&mut self.inner)
+    }
+
+    /// Take the inner cache, even if disabled
+    fn take_always(self) -> TxIndexCache {
```

Comment:
Instead of take_always(), we can just use the Option as returned value and then do `or_default()` and the map will be empty.